allennlp/modules/span_extractors/endpoint_span_extractor.py

```diff
@@ -86,15 +101,40 @@ def forward(self,
             span_starts = span_starts * span_indices_mask
             span_ends = span_ends * span_indices_mask
 
-        start_embeddings = batched_index_select(sequence_tensor, span_starts)
-        end_embeddings = batched_index_select(sequence_tensor, span_ends)
+        if not self._use_exclusive_start_indices:
+            start_embeddings = util.batched_index_select(sequence_tensor, span_starts)
+            end_embeddings = util.batched_index_select(sequence_tensor, span_ends)
 
-        combined_tensors = combine_tensors(self._combination, [start_embeddings, end_embeddings])
+        else:
+            # We want `exclusive` span starts, so we remove 1 from the forward span starts
+            # as the AllenNLP ``SpanField`` is inclusive.
+            # shape (batch_size, num_spans)
+            exclusive_span_starts = span_starts - 1
+            # shape (batch_size, num_spans, 1)
+            start_sentinel_mask = (exclusive_span_starts == -1).long().unsqueeze(-1)
+            exclusive_span_starts = exclusive_span_starts * (1 - start_sentinel_mask.squeeze(-1))
+
+            # We'll check the indices here at runtime, because it's difficult to debug
```

Comment:
When will this happen?  Seems like the right place to handle this is in tests?  Or replace the `== -1` above with `< 0`?  That is, checks that are aimed at bugs belong in tests; checks that are aimed at invalid inputs on user-facing functions belong as asserts in the code.  Seems like this is aiming at a bug, not at potentially invalid inputs.

Reply:
I've always been wary of asserts over raising `RuntimeError` or `ValueError`s, because you can pass `-O or -OO` flags to python which cause asserts to not be checked. Why you would want this is beyond me, but ever since reading that i've always used exceptions.

Reply:
Sorry, I didn't mean to say this should be an assert, and when I said "assert", I was including exceptions like this.  I was suggesting that this belongs in a test, not as a runtime check.

Reply:
To be more clear: I don't see how you can get into this state.  It looks like this is checking for a bug in the logic above.  Bug checks belong in tests.

Reply:
It's purpose is that if you did accidentally pass incorrect indices to this function, it wouldn't necessarily crash (because the indices get shifted by the batch dimension when we flatten to do the `index_select`, so if they are negative they won't necessarily stay negative) and it would be really, really hard to find.

Reply:
I see, ok.  In that case, would it make more sense to put this at the top of the function, and check it on `span_starts`?

Reply:
no, because you can pass a mask to this function, and we need to mask out the indices which correspond to using the start sentinel (which have an exclusive index of -1) before we check.

Reply:
But those are the same indices where you have a start index of 0.  That's the part where it looks like you're checking for a bug in your code, not for invalid input.

Reply:
Ok - I see what you mean - but I think that makes the semantics of what I am checking much more confusing. It's not directly a case of checking user input, because I have to mask it first to do the check, so to me it seems clearer to have the check closer to where it would break. It's also the same as in the BidirectionalEndpointSpanExtractor.